book-content/chapters/12-the-weird-parts.md

```diff
@@ -0,0 +1,1653 @@
+  releaseYear: album.releaseYear + 1,
+  strangeProperty: "This is strange",
+}));
+```
+
+Now, our `newAlbums` array will have an excess `strangeProperty` property on each `Album` object, without TypeScript even knowing about it. It thinks that the return type of the function is `Album[]`, but it's actually `(Album & { strangeProperty: string })[]`.
+
+The way we'd get this 'working' is to add a return type annotation to our inline function:
+
+```ts twoslash
+// @errors: 2353
+type Album = {
+  title: string;
+  releaseYear: number;
+};
+
+const remapAlbums = (albums: Album[], remap: (album: Album) => Album) => {
+  return albums.map(remap);
+};
+
+const albums: Album[] = [
+  { title: "Rubber Soul", releaseYear: 1965 },
+  { title: "Revolver", releaseYear: 1966 },
+  { title: "Sgt. Pepper's Lonely Hearts Club Band", releaseYear: 1967 },
+];
+
+// ---cut---
+const newAlbums = remapAlbums(
+  albums,
+  (album): Album => ({
+    ...album,
+    releaseYear: album.releaseYear + 1,
+    strangeProperty: "This is strange",
+  }),
+);
+```
+
+This will cause TypeScript to show an error for the excess `strangeProperty` property.
+
+This works because in this situation, we're comparing an inline object (the value we're returning) directly to a type. TypeScript will check for excess properties in this case.
+
+Without a return type annotation, TypeScript ends up trying to compare two functions, and it doesn't really mind if a function returns too many properties.
+
+### Open vs Closed Object Types
+
+TypeScript, by default, treats all objects as _open_. At any time, it expects that other properties might be present on objects.
+
+Other languages, like Flow, treat objects as _closed_ by default. Flow is Meta's internal type system, and by default requires objects to be exact (their term for 'closed').
+
+```js
+function method(obj: { foo: string }) {
+  /* ... */
+}
+
+method({ foo: "test", bar: 42 }); // Error!
+```
+
+You can opt in to open (or inexact) objects in Flow with a `...` syntax:
+
+```js
+function method(obj: { foo: string, ... }) {
+  /* ... */
+}
+
+method({ foo: "test", bar: 42 }); // No more error!
+```
+
+But Flow recommends you use closed objects by default. They think that, especially when working with spread operators, it's better to err on the side of caution.
+
+### Why Does TypeScript Treat Objects As Open?
+
+Open objects more closely reflect how JavaScript actually works. Any type system for JavaScript - a very dynamic language - has to be relatively cautious about how 'safe' it can truly be.
+
+So, TypeScript's decision to treat objects as open by default is a reflection of the language it's trying to type. It also more closely reflects how objects work in other languages.
+
+The issue is that the excess properties warning can often make you think TypeScript uses closed objects.
+
+But really, the excess properties warning is more like a courtesy. It's only used in cases where the object can't be modified elsewhere.
```

Comment:
+ it's a reason why we can have renaming suggestions in those errors like:
> Object literal may only specify known properties, but 'timeOut' does not exist in type '{ timeout?: number | undefined; a: number; }'. Did you mean to write 'timeout'?(2561)


Without the excess property checks (and with open objects) this wouldn't be an error so the typo would slip through. But, of course, like pointed out above - it still can slip through quite easily... given how this only is performed on inline object literals.